docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,197 @@
+A user can define additional `PodAntiAffinity`s but can't prevent the operator from creating the default `PodAntiAffinity`.
+The only exception is the special fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+In the design we will use the following terms:
+
+- `physical fault domain`: This refers to a fault domain that is available in Kubernetes e.g. a machine running the Kubernetes worker or a higher level fault domain like a rack, data hall etc.
+  This will reflect an underlying physical fault domain.
+- `logical fault domain`: The logical fault domain is not something that refers directly to a physical fault domain and can also be spread across multiple physical fault domains.
+  The idea is to have a way to group some processes of a FoundationDB cluster together with the same zone ID, this zone ID will be a logical value like "zone-1".
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we will use a prefix for the `BinPack` mode and a value defining in which zone the process should be running.
+We will also set the `FDB_ZONE_ID` environment variable on the Pod level, the default value of this field is currently the hostname.
+In addition to that we change the `PodAntiAffinity` rule.
+The following would be an example of the `affinity` term for a Pod.
+The example assumes that process should be running in zone `0` and we have set `BinPackAll`:
+
+```yaml
+  affinity:
+    # Affinity to schedule Pods with the same distribution key together
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+              foundationdb.org/fdb-cluster-name: example-cluster
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      # The required statement is per default only allowed on the hostname:
+      # https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+            - key: "foundationdb.org/fdb-cluster-name"
+              operator: In
+              values:
+              - "example-cluster"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across physical fault domains.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
```

Comment:
Do we have use cases defined for different bin packing strategies? Today we're always bin packing by class, and I'm not clear on the advantages of different strategies.

Reply:
Conclusion from discussion: We don't need different bin packing options with the current state of the design.